src/Npgsql/NpgsqlTypes/NpgsqlInterval.cs

```diff
@@ -0,0 +1,323 @@
+﻿using System;
+
+namespace NpgsqlTypes
+{
+    /// <summary>
+    /// Represents a time interval.
+    /// </summary>
+    public readonly struct NpgsqlInterval : IEquatable<NpgsqlInterval>, IComparable<NpgsqlInterval>
+    {
+        internal long Microseconds { get; }
+        internal int Days { get; }
+        internal int Months { get; }
+
+        internal NpgsqlInterval(long microseconds, int days, int months) =>
+            (Microseconds, Days, Months) = (microseconds, days, months);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NpgsqlInterval"/>
+        /// structure to the specified year, month, day, hour, minute, second, millisecond, and microsecond.
+        /// </summary>
+        /// <param name="year">Number of years.</param>
+        /// <param name="month">Number of months.</param>
+        /// <param name="day">Number of days.</param>
+        /// <param name="hour">Number of hours.</param>
+        /// <param name="minute">Number of minutes.</param>
+        /// <param name="second">Number of seconds.</param>
+        /// <param name="millisecond">Number of milliseconds.</param>
+        /// <param name="microsecond">Number of microseconds.</param>
+        public NpgsqlInterval(int year, int month, int day, int hour, int minute, int second = 0, int millisecond = 0, int microsecond = 0)
```

Comment:
Is it okay to have only one constructor like this or multiple are better? Since all parameters have the same type, I don't think it would be helpful, but replacing them by methods like `WithYear(int year, int month = 0, ...)` might solve the problem.

Reply:
Sorry, missed this message somehow...

I think the one constructor is probably better... As you say, if all parameters have the same type, it's very easy to get confused and use the wrong values etc. We could *maybe* consider having a constructor for time-only intervals (i.e. without year/month/day) since that seems to be a common scenario, but the two constructors would still probably conflict - maybe better leave it at one.

I don't think it's worth going into a builder-type API with WithYear right now (though NodaTime's Period does have that) - we can always add that later if there's user request for it.

One small note: I'd leave `second` without a default value (mandatory), while indeed having default 0 for sub-second values. This seems consistent with DateTime and other types.



> public NpgsqlInterval(int year, int month, int day, int hour, int minute, int second, int millisecond = 0, int microsecond = 0)
> public NpgsqlInterval(int hour, int minute, int second, int millisecond = 0, int microsecond = 0)